File: test/test_editor_ops.js

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PROJECT_PATH = path.resolve(__dirname, "../example_project");

async function main() {
    console.log("=== Starting Godot MCP Editor Operations Test ===");

    const transport = new StdioClientTransport({
        command: "node",
        args: [path.join(__dirname, "../src/index.js")]
    });

    const client = new Client({ name: "test-client", version: "1.0.0" }, { capabilities: {} });

    try {
        await client.connect(transport);
        console.log("[Test] Connected to MCP server.");

        // 1. Launch Instance
        console.log("[Test] Launching Godot...");
        await client.callTool({
            name: "godot_launch",
            arguments: {
                project_path: PROJECT_PATH,
                wait_for_ready: true,
                headless: true // Run editor headless for this test if possible? 
                // Note: godot_launch headless arg usually applies to game, but for editor tools we need editor.
                // We'll just launch normally.
            }
        });
        console.log("[Test] Godot Launched.");

        // 2. Project Settings
        console.log("[Test] Listing Project Settings...");
        await client.callTool({ name: "godot_project_settings_list", arguments: { prefix: "application/config" } });

        console.log("[Test] Getting specific setting...");
        await client.callTool({ name: "godot_project_setting_get", arguments: { setting_name: "application/config/name" } });

        console.log("[Test] Setting project setting...");
        // Use a safe setting or a custom one
        await client.callTool({
            name: "godot_project_setting_set",
            arguments: { setting_name: "application/config/test_mcp_prop", value: "test_value" }
        });

        // 3. Input Actions
        console.log("[Test] Listing Input Actions...");
        await client.callTool({ name: "godot_input_actions_list", arguments: {} });

        // 4. Scene & Node Operations
        console.log("[Test] Getting Current Scene...");
        const scene = await client.callTool({ name: "godot_scene_current", arguments: {} });
        const sceneInfo = JSON.parse(scene.content[0].text);
        const rootPath = sceneInfo.path; // e.g. res://scenes/test_framework.tscn (this returns file path, not node path)
        // Wait, godot_scene_current returns metadata about the OPEN scene tab.
        // We need the scene TREE path for node operations. Usually root is /root/Name

        console.log("[Test] Getting Scene Tree to find root node path...");
        const tree = await client.callTool({ name: "godot_scene_get_tree", arguments: {} });
        const treeData = JSON.parse(tree.content[0].text);
        const rootNodeName = treeData.root.name; // e.g. "TestRoot"
        const rootNodePath = "/root/" + rootNodeName; // Actually in Editor it's usually just the name if it's the edited scene root? 
        // In EditorPlugin logic, the edited scene root is usually relative to the Scene Dock.
        // Let's assume we can interact with children.

        console.log("[Test] Creating Node...");
        const nodeName = "MCPTestNode";
        await client.callTool({
            name: "godot_node_create",
            arguments: {
                parent_path: ".", // "." usually means scene root in my implementation? Need to verify. 
                // If implementation uses `get_node(parent_path)`, "." works.
                node_type: "Node",
                node_name: nodeName
            }
        });

        console.log("[Test] Setting Node Property...");
        await client.callTool({
            name: "godot_node_set_property",
            arguments: {
                node_path: nodeName, // Relative path from root?
                property: "name",
                value: "MCPTestNodeRenamed"
            }
        });

        console.log("[Test] Getting Node Info...");
        await client.callTool({ name: "godot_node_info", arguments: { node_path: "MCPTestNodeRenamed" } });

        console.log("[Test] Getting Node Properties...");
        await client.callTool({ name: "godot_node_properties", arguments: { node_path: "MCPTestNodeRenamed" } });

        // 5. Scripts
        console.log("[Test] Getting existing script...");
        // Use an existing node from the project
        // root node has script?
        await client.callTool({ name: "godot_script_get", arguments: { node_path: "." } }); // Root

        // godot_script_source removed (agents can read files directly)

        // 6. Cleanup Node
        console.log("[Test] Deleting Node...");
        await client.callTool({ name: "godot_node_delete", arguments: { node_path: "MCPTestNodeRenamed" } });

        // 7. Editor Output
        console.log("[Test] Reading Editor Output...");
        await client.callTool({ name: "godot_editor_output", arguments: {} });

        console.log("[Test] All Editor Operations Completed Successfully.");

    } catch (e) {
        console.error("[Test] Error:", e);
        process.exit(1);
    } finally {
        console.log("[Test] Terminating...");
        try {
            await client.callTool({ name: "godot_terminate", arguments: { instance_id: PROJECT_PATH } });
        } catch (e) { }
        process.exit(0);
    }
}

main();

File: test/test_editor_ops.py
#!/usr/bin/env python3
"""Editor operations test: launches a real Godot editor and exercises editor tools over MCP."""

import asyncio
import json
import sys
from pathlib import Path

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

REPO_ROOT = Path(__file__).resolve().parents[1]
PROJECT_PATH = str(REPO_ROOT / "example_project")
SERVER_CMD = str(REPO_ROOT / ".venv" / "bin" / "godot-mcp-server")


def text_of(result):
    return result.content[0].text


async def call(session, name, args=None, expect_ok=True):
    result = await session.call_tool(name, args or {})
    if expect_ok and result.isError:
        raise RuntimeError(f"{name} failed: {text_of(result)}")
    return result


async def main():
    print("=== Starting Godot MCP Editor Operations Test ===")
    params = StdioServerParameters(command=SERVER_CMD)

    async with stdio_client(params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()
            print("[Test] Connected to MCP server.")

            try:
                print("[Test] Launching Godot...")
                await call(session, "godot_launch", {
                    "project_path": PROJECT_PATH,
                    "wait_for_ready": True,
                })
                print("[Test] Godot Launched.")

                print("[Test] Listing Project Settings...")
                await call(session, "godot_project_settings_list",
                           {"prefix": "application/config"})

                print("[Test] Getting specific setting...")
                result = await call(session, "godot_project_setting_get",
                                    {"setting_name": "application/config/name"})
                print("[Test] Setting value:", text_of(result)[:120])

                print("[Test] Setting project setting...")
                await call(session, "godot_project_setting_set", {
                    "setting_name": "application/config/test_mcp_prop",
                    "value": "test_value",
                })

                print("[Test] Listing Input Actions...")
                await call(session, "godot_input_actions_list")

                print("[Test] Getting Current Scene...")
                scene = await call(session, "godot_scene_current")
                print("[Test] Current Scene:", text_of(scene)[:200])

                print("[Test] Getting Scene Tree...")
                tree = await call(session, "godot_scene_get_tree")
                tree_data = json.loads(text_of(tree))
                print("[Test] Scene root:", tree_data.get("root", {}).get("name"))

                print("[Test] Creating Node...")
                node_name = "MCPTestNode"
                await call(session, "godot_node_create", {
                    "parent_path": ".",
                    "node_type": "Node",
                    "node_name": node_name,
                })

                print("[Test] Setting Node Property...")
                await call(session, "godot_node_set_property", {
                    "node_path": node_name,
                    "property": "name",
                    "value": "MCPTestNodeRenamed",
                })

                print("[Test] Getting Node Info...")
                await call(session, "godot_node_info", {"node_path": "MCPTestNodeRenamed"})

                print("[Test] Getting Node Properties...")
                await call(session, "godot_node_properties", {"node_path": "MCPTestNodeRenamed"})

                print("[Test] Getting existing script...")
                await call(session, "godot_script_get", {"node_path": "."})

                print("[Test] Deleting Node...")
                await call(session, "godot_node_delete", {"node_path": "MCPTestNodeRenamed"})

                print("[Test] Reading Editor Output...")
                await call(session, "godot_editor_output")

                print("[Test] Listing dialogs (expect none)...")
                dialogs = await call(session, "godot_editor_dialogs")
                assert json.loads(text_of(dialogs))["count"] == 0, text_of(dialogs)

                print("[Test] play 'main' with bad scene fails fast (no modal)...")
                await call(session, "godot_project_setting_set", {
                    "setting_name": "application/run/main_scene", "value": "res://nope.tscn"})
                bad_play = await session.call_tool("godot_game_play", {"scene_mode": "main"})
                assert bad_play.isError and "does not exist" in text_of(bad_play), text_of(bad_play)
                # Guard should mean no modal popped
                dialogs2 = await call(session, "godot_editor_dialogs")
                assert json.loads(text_of(dialogs2))["count"] == 0, \
                    f"unexpected modal: {text_of(dialogs2)}"
                await call(session, "godot_project_setting_set", {
                    "setting_name": "application/run/main_scene",
                    "value": "res://scenes/test_scene.tscn"})
                print("[Test] Play failed fast with no modal. ✔")

                print("[Test] All Editor Operations Completed Successfully.")
            finally:
                print("[Test] Terminating...")
                try:
                    await call(session, "godot_terminate",
                               {"instance_id": PROJECT_PATH}, expect_ok=False)
                except Exception:
                    pass


if __name__ == "__main__":
    asyncio.run(main())
